modules/http4s/test/src/smithy4s/http4s/NullsAndDefaultEncodingSuite.scala

```diff
@@ -0,0 +1,235 @@
+import org.typelevel.ci.CIString
+import org.typelevel.ci._
+import org.http4s.circe.CirceInstances
+import smithy4s.Document
+import org.http4s.client.Client
+import smithy4s.example.OperationInput
+import cats.effect.kernel.Deferred
+
+object NullsAndDefaultEncodingSuite extends SimpleIOSuite with CirceInstances {
+
+  test("routes - explicit defaults encoding = false") {
+    runServerTest(explicitDefaults = false).map { response =>
+      assert.same(
+        Map(ci"required-header-with-default" -> "required-header-with-default"),
+        response.headers
+      ) &&
+      assert.same(
+        Json.obj(
+          "requiredWithDefault" -> Json.fromString("required-default")
+        ),
+        response.body
+      )
+    }
+  }
+
+  test("routes - explicit defaults encoding = true") {
+    runServerTest(explicitDefaults = true).map { response =>
+      assert.same(
+        Map(ci"required-header-with-default" -> "required-header-with-default"),
+        response.headers
+      ) &&
+      assert.same(
+        Json.obj(
+          "requiredWithDefault" -> Json.fromString("required-default"),
+          "optionalWithDefault" -> Json.fromString("optional-default"),
+          "optional" -> Json.Null
+        ),
+        response.body
+      )
+    }
+  }
+
+  test("client - explicit defaults encoding = false") {
+    runClientTest(explicitDefaults = false, OperationInput())
+      .map { request =>
+        assert.same(
+          Map(
+            ci"required-header-with-default" -> "required-header-with-default"
+          ),
+          request.headers
+        ) && assert.same(
+          Map.empty,
+          request.query
+        ) && assert.same(
+          Json.obj(
+            "requiredWithDefault" -> Json.fromString("required-default")
+          ),
+          request.body
+        )
+      }
+  }
+
+  test("client - explicit defaults encoding = true") {
+    runClientTest(explicitDefaults = true, OperationInput())
+      .map { request =>
+        assert.same(
+          Map(
+            ci"optional-header-with-default" -> "optional-header-with-default",
+            ci"required-header-with-default" -> "required-header-with-default",
+            ci"optional-header" -> ""
+          ),
+          request.headers
+        ) && assert.same(
+          Map(
+            "optional-query-with-default" -> "optional-query-with-default",
+            "required-query-with-default" -> "required-query-with-default",
+            "optional-query" -> ""
+          ),
+          request.query
+        ) && assert.same(
+          List("required-label-with-default"),
+          request.labels
+        ) && assert.same(
+          Json.obj(
+            "optional" -> Json.Null,
+            "optionalWithDefault" -> Json.fromString("optional-default"),
+            "requiredWithDefault" -> Json.fromString("required-default")
+          ),
+          request.body
+        )
+      }
+  }
+
+  test("document encoder - all default") {
```

Comment:
**Question**: Do we need a flag to control the `Document.Encoder` behavior?

Reply:
imo yes

Reply:
Considering it's in `internals`, we deem it's okay to break. If you're a third party who depends on what's in an `internals` package, it's your funeral. 

Moreover, the reason we've not been ultra strict wrt private/public is because of a drastic lack of tooling that would help us do so. We'd like to build such tooling (probably by means of scalafix)

But I think it'd be totally possible to retain binary compatibility by means of secondary constructors.

Reply:
ah, I didn't realize it was in `internals`. Alright then